omnigibson/prims/rigid_prim.py

```diff
@@ -315,69 +317,116 @@ def get_angular_velocity(self, clone=True):
         """
         return self._rigid_prim_view.get_angular_velocities(clone=clone)[0]
 
-    def set_position_orientation(self, position=None, orientation=None):
+    def set_position_orientation(
+        self, position=None, orientation=None, frame: Literal["world", "parent", "scene"] = "world"
+    ):
+        """
+        Set the position and orientation of XForm Prim.
+
+        Args:
+            position (None or 3-array): The position to set the object to. If None, the position is not changed.
+            orientation (None or 4-array): The orientation to set the object to. If None, the orientation is not changed.
+            frame (Literal): The frame in which to set the position and orientation. Defaults to world. parent frame
+            set position relative to the object parent. scene frame set position relative to the scene.
+        """
+
+        assert frame in ["world", "parent", "scene"], f"Invalid frame '{frame}'. Must be 'world', 'parent', or 'scene'."
+
+        # compute the pose in the desired frame
+        position, orientation = T.compute_desired_pose_in_frame(self, position, orientation, frame=frame)
+
         # Invalidate kinematic-only object pose caches when new pose is set
         if self.kinematic_only:
             self.clear_kinematic_only_cache()
-        if position is not None:
-            position = th.asarray(position)[None, :]
-        if orientation is not None:
-            assert math.isclose(
-                th.norm(orientation), 1, abs_tol=1e-3
-            ), f"{self.prim_path} desired orientation {orientation} is not a unit quaternion."
-            orientation = th.asarray(orientation)[None, [3, 0, 1, 2]]
-        self._rigid_prim_view.set_world_poses(positions=position, orientations=orientation)
+
+        position = th.asarray(position)[None, :]
+        assert math.isclose(
+            th.norm(orientation).item(), 1, abs_tol=1e-3
+        ), f"{self.prim_path} desired orientation {orientation} is not a unit quaternion."
+        orientation = th.asarray(orientation)[None, [3, 0, 1, 2]]
```

Comment:
Do this right at the beginning right after the `assert frame in ...`. Make sure you manually get the pos/orn from the object if not provided.